v2/lints/cabf_br/lint_ocsp_id_pkix_ocsp_nocheck_ext_not_included.go

```diff
@@ -0,0 +1,58 @@
+/*
+ * ZLint Copyright 2020 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package cabf_br
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v2/lint"
+	"github.com/zmap/zlint/v2/util"
+)
+
+type OCSPIDPKIXOCSPNocheckExtNotIncluded struct{}
+
+func (l *OCSPIDPKIXOCSPNocheckExtNotIncluded) Initialize() error {
+	return nil
+}
+
+func (l *OCSPIDPKIXOCSPNocheckExtNotIncluded) CheckApplies(c *x509.Certificate) bool {
+	return util.HasEKU(c, x509.ExtKeyUsageOcspSigning)
+}
+
+func (l *OCSPIDPKIXOCSPNocheckExtNotIncluded) Execute(c *x509.Certificate) *lint.LintResult {
+
+	// The OCSPNoCheckOID is included in the certficate this is a clear pass (at least for this lint)
+	if util.IsExtInCert(c, util.OscpNoCheckOID) {
+		return &lint.LintResult{Status: lint.Pass}
+	}
+
+	if util.IsServerAuthCert(c) {
+		// If the certificate is a TLS certificate, it is clear, that the BRGs apply and we have an ERROR
+		return &lint.LintResult{Status: lint.Error}
+	}
+
+	// If the certificate is not a TLS certificate, the BRGs only apply, if the parent certificate has the Server Auth EKU or could possibly issue a certificate that has the Server EKU
+	return &lint.LintResult{Status: lint.Warn, Details: "If the parent (issuing) certificate contains the Server Auth EKU or no EKU at all this is an ERROR and not a WARN"}
```

Comment:
However, I'm not entirely certain we need to handle this case, and I'd be curious for feedback from other @zmap/zlint-maintainers .

Today, the CABF BR lints [are](https://github.com/zmap/zlint/blob/aa4e2619db00dbdea768e9afa70262f0d6af3417/v2/util/ca.go#L47) [run](https://github.com/zmap/zlint/blob/aa4e2619db00dbdea768e9afa70262f0d6af3417/v2/lint/base.go#L86):
- If the cert has no EKUs
- If the cert has >=1 EKU, and contains `id-kp-anyExtendedKeyUsage`
- If the cert has >=1 EKU, and contains `id-kp-serverAuth`

That means if you have a hierarchy, where you have
- Root CA
- Intermediate CA 1 (only has `id-kp-emailProtection`)
- Intermediate CA 2 (no EKU extension)
- Leaf Cert (`id-kp-emailProtection`)

Then we run all of the BR lints against `Intermediate CA 2`, which very likely will have some that create warnings/failures, even though it's beneath a technically-constrained sub-CA.

This PR updates the BR checks to *also* run if:
- The cert has >=1 EKU, and contains `id-kp-ocspSigning`

And this logic here is trying to handle the scenario of:
- Root CA
- Intermediate CA 1 (only has `id-kp-emailProtection`)
- OCSP Signer (has `id-kp-ocspSigning`)

Trying to handle "OCSP Signer" seems odd, because we don't handle "Intermediate CA 2" any specially. But I guess the argument is that "Intermediate CA 2" is generally unlikely from publicly trusted hierarchies, because different root programs (e.g. Microsoft) have long required that intermediates have EKUs, so while "Intermediate CA 2" is unlikely/rare, "OCSP Signer" is far more likely/common.

I have a _slight_ preference for making this consistently an error, consistent with how the other BR checks apply in the absence of the EKU, and that effectively means "assume OCSP signer certs are subject to the BRs", but I'm curious for other thoughts, especially from @cardonator (to understand how CAs with complex hierarchies have integrated and dealt with the existing false-positives)

Reply:
Regarding false positives, we do have a list of lints that tend to fail for not legitimate reasons that we review on a regular basis but don't prevent us from issuing a new ICA (or EE cert). The lints are still logged for review but they don't prevent us from issuing. Obviously, recent iterations of our system don't even allow us to create the hierarchy you mentioned above 😄 but there are several other scenarios in which a lint would fail for a not legitimate reason, particularly on ICAs and even more particularly when linting prior to the actual signing taking place.

Reply:
@cardonator would the current distinction with `warning` and `error` be okay for you?

Reply:
I have split the one lint in two. There is one lint for `warning` and one for `error` now. I have also created test cases for all possible (32) combinations of the EKU OCSPSigning, serverAuth, emailProtection and extendedKeyUsage with OCSPNoCheck.